crates/runtime/src/dataconnector/github/rate_limit.rs

```diff
@@ -0,0 +1,233 @@
+use chrono::{DateTime, TimeZone, Utc};
+use reqwest::header::HeaderMap;
+use std::{
+    sync::Arc,
+    time::{Duration, SystemTime},
+};
+use tokio::sync::RwLock;
+use tracing::{debug, warn};
+
+const GITHUB_MAX_REQUESTS_PER_HOUR: u32 = 5000;
+
+#[derive(Debug)]
+pub struct RateLimiter {
+    // Track API response headers rate limits
+    api_limit: Arc<RwLock<Option<RateLimitInfo>>>,
+    // Track our own rate limiting
+    requests_made: u32,
+    start_time: SystemTime,
+}
+
+#[derive(Debug, Clone)]
+pub struct RateLimitInfo {
+    pub limit: i32,
+    pub remaining: i32,
+    pub used: i32,
+    pub reset_time: DateTime<Utc>,
+    pub resource: String,
+}
+
+// See https://docs.github.com/en/graphql/overview/rate-limits-and-node-limits-for-the-graphql-api#checking-the-status-of-your-primary-rate-limit
+impl RateLimitInfo {
+    pub fn from_headers(headers: &HeaderMap) -> Option<Self> {
+        let limit = headers
+            .get("x-ratelimit-limit")?
+            .to_str()
+            .ok()?
+            .parse::<i32>()
+            .ok()?;
+        let remaining = headers
+            .get("x-ratelimit-remaining")?
+            .to_str()
+            .ok()?
+            .parse::<i32>()
+            .ok()?;
+        let used = headers
+            .get("x-ratelimit-used")?
+            .to_str()
+            .ok()?
+            .parse::<i32>()
+            .ok()?;
+        let reset = headers
+            .get("x-ratelimit-reset")?
+            .to_str()
+            .ok()?
+            .parse::<i64>()
+            .ok()?;
+        let resource = headers
+            .get("x-ratelimit-resource")?
+            .to_str()
+            .ok()?
+            .to_string();
+
+        let reset_time = Utc.timestamp_opt(reset, 0).unwrap();
+
+        Some(Self {
+            limit,
+            remaining,
+            used,
+            reset_time,
+            resource,
+        })
+    }
+}
+
+impl RateLimiter {
+    pub fn new() -> Self {
+        Self {
+            api_limit: Arc::new(RwLock::new(None)),
+            requests_made: 0,
+            start_time: SystemTime::now(),
+        }
+    }
+
+    pub async fn update_from_headers(&self, headers: &HeaderMap) {
+        if let Some(rate_limit) = RateLimitInfo::from_headers(headers) {
+            let mut api_limit = self.api_limit.write().await;
+            *api_limit = Some(rate_limit);
+        }
+    }
+
+    pub async fn check_rate_limit(&mut self) -> Result<(), String> {
+        // First check our internal hourly limit
+        if let Ok(elapsed) = self.start_time.elapsed() {
+            if elapsed >= Duration::from_secs(3600) {
```

Comment:
Resetting the rate limit at the hour boundary won't correctly handle bursts of requests made right before the reset.

Should we consider using a dedicated library instead of our own implementation?
https://crates.io/keywords/rate-limit

Reply:
We should use the values from the headers directly instead of trying to implement our own rate limiter logic on top of GitHub's. The headers from GitHub tell us all of this information.